### ISSUE TITLE
Sidecar performance degrades over time in Istio: 1.2.3 

### ISSUE THREAD

**Issue by nokiabaitanik** [istio/istio#16370]
(NOTE: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities/>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>
)

**Bug description**

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ X ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior : After side car Injected , if performance drops , to 50%. but main problem ins it is degrading over time. It should NOT degrade over time. *** 

**Steps to reproduce the bug**
Without side car :
1. client simulator (vegeta) to send GET/POST request -> application pod using pod service IP
In statistics we can see it is able to reach 8k tps for 8 hours run . it was stable for 8hours.
2. Side car attached to the pod. (we have mTLS disabled)
client simulator (vegeta) to send GET/POST request -> application pod using pod service IP.
In statistics we can see it was reaching upto 4k . BUT problem is that it is degrading over time.
it 8 hours end it was around 500

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
istio 1.2.3

**How was Istio installed?**
we installed as  kubernetes helm chart. 

**Environment where bug was observed (cloud vendor, OS, etc)**
Environment:
We are using kubernetes cluster.
Deployment pod were given CPU: 4., Memory: 4Gi
Application is using 2Gi Memory.
So side car has 2 Gb memory to use.
For istio side car :
resources:
limits:
cpu: "2"
memory: 1Gi
requests:
cpu: 100m
memory: 128Mi

I had raised the following ticket . but not getting any update there .. 
https://github.com/istio/istio/issues/15981 

Details are there in the other ticket .

**Comment by howardjohn** [comment]
What if you set the requests on the sidecar to the same as the limit? What may happen is when you start  the test, the sidecar has access to the full 2 CPUs it is limited to, but as it goes on the node becomes under load, and eventually the sidecar gets throttled and is limited before 2 cores

**Comment by nokiabaitanik** [comment]
We have tried that . for sidecar :

resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: "2"
        memory: 1Gi

Still after sometime it starts to go down.

**Comment by nokiabaitanik** [comment]
Can you in your env check for longer run (may be 12/24 hours ) whether tps is degrading ?

**Comment by mandarjog** [comment]
How many clients / client connections were in use during the test?
How many server replicas were there?  Was the traffic going they ingress?
Is sidecar attached to client and server or just server ?

**Comment by nokiabaitanik** [comment]
Client is only one. vegeta simulator. Which is creating http2 connections to the application pod.

rest client simulator(vegeta/hey) -> application service ip (side car is attached)

2. How many server replicas were there? Was the traffic going they ingress? - 
    server means the application pod is only one. We are trying to figure it out max capacity for a single pod. then according to requirement we are going to increase the number. 

We are not sending traffic through ingress gateway. First we are trying to figure out what is the performance drop if we only attach side car.
So that is why we are directly sending traffic to application service ip/port while side car is attached.

3. Is sidecar attached to client and server or just server ? -- side car attached to only server ( application pod) . Client is just a open source simulator . outside the kubernetes cluster.


Thing is with the same setup without side car
simulator -> application service IP (kubectl get svc ) when side car is NOT attached . tps (max 8k or9k )is stable for 24 hours.

same env , just only side car is attached .
simulator -> application service IP (kubectl get svc ) when side car is attached . tps (max 3k-4k )is NOT stable for long duration. after some time it starts degrading.

if we are sending 4k (side car attached ) it will start degrade within 5-10 minutes. 
if we sending only 500 (side car attached) it will start degrading after around 2 hours.

Screen shots are attached in https://github.com/istio/istio/issues/15981

**Comment by mandarjog** [comment]
Thanks. You mentioned client creates h2 connections, can you point me to your kubernetes config? Is your server also running h2 or is it running dual stack?

**Comment by mandarjog** [comment]
@carolynhu

**Comment by nokiabaitanik** [comment]
yes.. dual stack.. it can support http1.1 or http2..  It does not support h2c (upgrade request) .
kubernetes config means which configuration u need ?

The resources for side car and application I have mentioned above.  Please let me know which config u need.

**Comment by mandarjog** [comment]
If your server side is capable of h2 then you should mark the port as http2 or grpc.
Otherwise it is being downgraded to http1.
I would like to see your Kuberentes service spect / deployment spec for the service under test.

**Comment by nokiabaitanik** [comment]
service spec:
spec:
  clusterIP: 10.254.108.255
  ports:
  - name: http
    port: 8080
    protocol: TCP
    targetPort: http
  - name: http-meters
    port: 8081
    protocol: TCP
    targetPort: http-meters
  - name: gogo
    port: 17000
    protocol: TCP
    targetPort: 17000
  selector:
    app.kubernetes.io/instance: xxxxx
    app.kubernetes.io/name: xxxxx
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}

**Comment by nokiabaitanik** [comment]
One more thing we observed today, that if we change the name ,
name: http -> to name:tcp .. it is not degrading much.  is that name matter in istio?

**Comment by nokiabaitanik** [comment]
we have tried http2... But it still degrades after some time.. 
If in http1 it was starting degrading after 5 minutes only ... but here it starts degrading after 8-10 minutes 


  
  ports:
  - name: http2
    port: 8080
    protocol: TCP
    targetPort: http2
  - name: http2-meters
    port: 8081
    protocol: TCP
    targetPort: http-meters
  - name: gogo
    port: 17000
    protocol: TCP
    targetPort: 17000
  selector:
    app.kubernetes.io/instance: csd-nscp
    app.kubernetes.io/name: scp
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}


But in case of tcp it is more or less stable. But with TCP we will not be able to get kiali visualization. 
Is it possible to get kali visualization in case of tcp

**Comment by nokiabaitanik** [comment]
Server images is not public . we are using centos 7.6. 
U need the entire config of the application service/deployment ? or istio charts?

**Comment by carolynhu** [comment]
@mandarjog sure, will try to reproduce this setup asap.

**Comment by nokiabaitanik** [comment]
Hi. Yes for all test telemetry was disabled. 
mixer.telemetry.enabled = false and mixer.policy.enabled = false ..

[appli_deployment.txt](https://github.com/istio/istio/files/3555493/appli_deployment.txt)
[config_dump.txt](https://github.com/istio/istio/files/3555494/config_dump.txt)
[appli_configmp.txt](https://github.com/istio/istio/files/3555495/appli_configmp.txt)

In the config_dump file .. "appli1" is the concerned component..

**Comment by nokiabaitanik** [comment]
we have found the issue .. it was with  http2 stack we were using. there was a memory leak. We have got a fix from that stack , now it is not degrading over time.  thanks for all the co operation and help...   
I will close the issue . Just one question .

if in the following scenario 
client -> service ip of the application -> http2 server/ outside world -- we are getting 8k tps .

then if we introduce all the ingress /egress then how much tps we should expect  ?? 
(mixer ,telemetry all enabled)
client -> ingress edge ip -> application with sidecar -> egress -> http2 server/ outside world ..
